### ISSUE TITLE
Unbalanced proximity analysis option

### ISSUE THREAD

**Issue by jimmymathews** [nadeemlab/SPT#20]
## Background
Before discussing the symmetric/non-symmetric distinction, first note that there is some tension here between:
1. *number* of cells or pairs
2. *distance* between cells

I say "tension" because it seems that a single number isn't enough to capture information of both (1) and (2).

I prefer to deal with this tension in what seems to be a standard way, namely to view the distance as a real-valued random variable of some sort, with a corresponding distribution D on R. To recover (1), one looks at values of the CDF of D, then multiplies as necessary by a total number of cells or pairs in order to recover absolute numbers as opposed to fractions. (It may be helpful to think of this last part as working with measure theory more generally, as opposed to probability theory, with the total "mass" not equal to 1).

The currently implemented "balanced" proximity workflow works as follows:

- Let C be the metric space of cell locations, and let P1 and P2 be 2 phenotypes considered as binary variables on C. Let P = (P1, P2) on C x C. The distance function d: C x C -> R can be considered a "random variable" on mass space C x C. We look at the distribution of ( d given P = (1, 1) ). That is, we restrict to the mass subspace defined by P = (1, 1) with uniform mass function equal to 1. The resulting restriction has a CDF-like function F whose plots display information about (1) and (2).

In the non-symmetric version we have more work to do:

- C, P, and d are as above. Now we regard (d given P = (1, 1) ) differently, as a family of functions on the second axis, parameterized by the value of the first axis in ( P1 = 1 subset of C). Each of these functions has its own CDF-like function, and what we do is average these functions to a single CDF-like function F.  Then "the average number of P2 within X of a given P1" is nothing but F(X).

As for the outcome pair testing, the sensible thing to do here is to regard the domain parameter X of F as a single parameter for a binary classifier using the value F(X) (recall that there is one F for each patient/case). We can either use manual/intuitive selection of X (30 microns), or the somewhat more rigorous AUC business that incorporates all values of X, which then leads to a best X for e.g. sum of specificity and sensitivity.

## Proposal
For this issue:
1. Refactor the phenotype proximity workflow so that the language it uses closely mirrors the description above.
2. Add a configuration variable, e.g. "balanced" vs. "unbalanced", which selects between the 2 behaviors described above.

**Comment by jimmymathews** [comment]
I think the balanced/unbalanced cases are more closely related than I previously thought.

**Claim**. *The average # of P2 within X of P1 is equal to (1 / # of P1) \* # of cell pairs of phenotype (P1, P2) of distance < X.*

**Proof**. Let S be the set of cell pairs of type (P1, P2) of distance < X. S is the disjoint union of S(c) = X-limited cell pairs with first entry equal to c, as c ranges over P1. So the size |S| can be computed as sum of |S(c)| over c in P1. The claim is this equation divided by |P1|.

It's the same basic trick as the proof of [Burnside's lemma](https://en.wikipedia.org/wiki/Burnside%27s_lemma). In the language of my first mesage above, the idea is that the sum of the 2 unnormalized CDFs of 2 respective sets of values is just the unnormalized CDF of the union of the 2 sets.

In other words, implementing the non-symmetric version is actually very easy: Just divide by |P1| !

This also means that the (P1, P2) averages are closely related to the (P2, P1) averages; just multiply by |P2| / |P1|.

Can someone check this logic?

**Comment by sanadeem** [comment]
Makes perfect sense to me. @rvanguri?

**Comment by jimmymathews** [comment]
Rami has requested that in the "balanced" case, we report cell pairs per unit cell area, while in the "unbalanced" case we report average neighbor numbers.

It seems that by nature the two cases suggest different normalizations. E.g. if neither endpoint of an edge in a graph is singled out, it doesn't make sense to use "number of neighbors" of the endpoint(s) as a unit.

This normalization behavior (and the main behavior of this issue) have been implemented: cf39a3d .

**Comment by jimmymathews** [comment]
What was actually implemented in the balanced case is normalization per unit compartmental area (on a per-FOV basis). This is the only place that dependence on regional/compartmental areas is present.
Later on it was requested that this dependence be removed. Normalization by the total cell area will be used instead, as suggested by the above comment.

**Comment by jimmymathews** [comment]
Dependence on compartmental areas file is removed: ce2f1db9e296bac4cce933cf2aa9efb8e2389ab3 . Unit and integration tests pass.
